### PR TITLE
FINERACT-1981: Interest recalculation progressive loan

### PR DESCRIPTION
**Pull request by somasorosdpc**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by adamsaghy** [comment]
@somasorosdpc Please make sure you are following the proper PR title format:
`FINERACT-XXXX: <short description>`

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 1]
```diff
@@ -757,7 +760,113 @@ private void allocateOverpayment(LoanTransaction loanTransaction, TransactionCtx
...
+                installment.updatePrincipal(rm.getPrincipalDue().getAmount());
+                installment.updateInterestCharged(rm.getInterestDue().getAmount());
+                installment.setRecalculatedInterestComponent(true);
+            }
+        });
+    }
+
     private void handleRepayment(LoanTransaction loanTransaction, TransactionCtx transactionCtx) {
+        if (transactionCtx instanceof ProgressiveTransactionCtx) {
+            log.info("handleRepayment called proper context.");
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: Please remove this. Its rathe belongs to `debug` level.

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 2]
```diff
@@ -757,7 +760,113 @@ private void allocateOverpayment(LoanTransaction loanTransaction, TransactionCtx
...
+            recalculateInterestForDate(loanTransaction.getTransactionDate(), (ProgressiveTransactionCtx) transactionCtx, false);
+        } else {
+            // TODO if we dont have a Progressive Context, we should somehow recreate it.
+            // We need to have a model which we can calculate overdue installments
+            // Should we be able to recreate model from installments without replaying all the transactions?
+            log.warn("Wrongly processed transaction context.");
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: Please remove this. This is not needed and wrong to warn for anything...

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 3]
```diff
@@ -757,7 +760,113 @@ private void allocateOverpayment(LoanTransaction loanTransaction, TransactionCtx
...
+    private List<LoanRepaymentScheduleInstallment> findOverdueInstallmentsBeforeDateSortedByInstallmentNumber(LocalDate currentDate,
+            ProgressiveTransactionCtx transactionCtx) {
+        final LocalDate fromDate = transactionCtx.getInstallments().get(0).getLoan().getApprovedOnDate();
+        return transactionCtx.getInstallments().stream().filter(installment -> !installment.getFromDate().isBefore(fromDate))
+                .filter(installment -> installment.getDueDate().isBefore(currentDate))
+                .filter(installment -> installment.isOverdueOn(currentDate))
+                .sorted(Comparator.comparing(LoanRepaymentScheduleInstallment::getInstallmentNumber)).toList();
+    }
+
+    private void recalculateInterestForDate(LocalDate currentDate, ProgressiveTransactionCtx ctx, boolean isLastRecalculation) {
```
Comment (adamsaghy) [inline fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:772]: I dont really understand this logic. Why do we need to recalculate interest after the full reprocessing of the transactions? Shouldnt we do it rather as part of the reprocessing? I mean when we are reprocessing the transactions thats the moment when something can go into overdue and interest must be recalculated hence it was not paid on time, so the next processed transaction might need to pay more...

**Comment by samr874** [comment]
@adamsaghy  are we using jira for maintainace of the application? Where will I get complete details about contribution and the tools used?

**Review by galovics — APPROVED** [review]

**Comment by adamsaghy** [comment]
> @adamsaghy are we using jira for maintainace of the application? Where will I get complete details about contribution and the tools used?

@samr874 https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103094318

File: fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 2 above)
Comment (somasorosdpc) [reply fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java]: removed

**Pull request merged**